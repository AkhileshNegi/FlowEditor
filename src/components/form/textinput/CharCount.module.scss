@import "variables.module.scss";

.count {
  text-align: right;
  padding: 5px 0;
}

.tooltip {
  position: relative;
  display: inline-block;
  font-size: 14px;
  min-width: 50px;
}

.fine_print {
  margin-top: 5px;
  font-size: 12px;
  line-height: 13px;
  color: lighten($gray, 20%);
}

.divider {
  padding-left: 3px;
  padding-right: 3px;
}

.tooltip .tooltip_text {
  visibility: hidden;
  min-height: 65px;
  min-width: 185px;
  background-color: white;
  color: $gray;
  text-align: left;
  padding: 12px;
  border-radius: 4px;
  font-size: 14px;
  line-height: 20px;
  -webkit-box-shadow: $shadow_down;
  box-shadow: $shadow_down;

  /* Position tooltip text */
  position: absolute;
  z-index: 1;
  bottom: 138%;
  left: -120%;

  /* Fade in tooltip */
  opacity: 0;
  -webkit-transition: opacity 0.4s;
  -o-transition: opacity 0.4s;
  transition: opacity 0.4s;
}

.tooltip .tooltip_text .unicode {
  left: -145%;
}

/* Show tooltip text when mouse hovers over the tooltip container */
.tooltip:hover .tooltip_text {
  visibility: visible;
  opacity: 1;
}
