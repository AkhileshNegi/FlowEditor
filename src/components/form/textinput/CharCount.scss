@import '../../../variables.scss';

.count {
    text-align: right;
    padding: 5px 0;
}

.tooltip {
    position: relative;
    display: inline-block;
    font-size: 14px;
    min-width: 50px;
}

.divider {
    padding-left: 3px;
    padding-right: 3px;
}

.tooltip .tooltip-text {
    visibility: hidden;
    min-height: 65px;
    min-width: 185px;
    background-color: white;
    color: $gray;
    text-align: left;
    padding: 12px;
    border-radius: 4px;
    font-size: 14px;
    line-height: 20px;
    -webkit-box-shadow: $shadow_down;
    box-shadow: $shadow_down;

    /* Position tooltip text */
    position: absolute;
    z-index: 1;
    bottom: 138%;
    left: -130%;

    /* Fade in tooltip */
    opacity: 0;
    -webkit-transition: opacity 0.4s;
    -o-transition: opacity 0.4s;
    transition: opacity 0.4s;
}

.tooltip .tooltip-text .unicode {
    left: -145%;
}

/* Show tooltip text when mouse hovers over the tooltip container */
.tooltip:hover .tooltip-text {
    visibility: visible;
    opacity: 1;
}
