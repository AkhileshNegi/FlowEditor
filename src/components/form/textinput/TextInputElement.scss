@import '~/variables.scss';

.option_list {
    overflow-y: auto;
    list-style-type: none;
    padding: 0;
    max-height: 300px;
    margin-bottom: 0px;
    margin-block-start: 0;
    margin-block-end: 0;
}

.option_summary {
    font-size: 12px;
    padding: 3px 0;
    font-weight: 300;
}

.option_signature {
    font-weight: 400;
    font-size: 13px;
    padding-bottom: 8px;
    padding-top: 3px;
    display: inline-block;
    padding-left: 3px;
}

.option {
    border-top: 1px solid transparent;
    border-bottom: 1px solid transparent;
    padding: 5px;
    font-size: 14px;
    &.selected {
        padding-top: 10px;
        padding-top: 10px;
        border-color: $blue_selected;
        background: $blue_selected;
        color: $text_color;
        font-weight: 500;
        &.first_option {
            border-top-color: transparent !important;
        }
    }
}

.fn_marker {
    display: inline-block;
    padding-right: 5px;
    font-style: italic;
}

.function_help {
    overflow: hidden;
    position: absolute;
    background: #fff;
    width: 250px;
    border-radius: 4px;
    -webkit-box-shadow: $shadow_down;
    box-shadow: $shadow_down;
    z-index: $z_completion;
}

.completion_container {
    overflow: hidden;
    position: absolute;
    background: #fff;
    margin-top: 20px;
    width: 250px;
    border-radius: 4px;
    -webkit-box-shadow: $shadow_down;
    box-shadow: $shadow_down;
    z-index: $z_completion;
}

.wrapper {
    position: relative;
}

.help {
    font-size: 10px;
    color: #999;
    padding: 4px;
    border-radius: 0 0 4px 4px;
    background-color: #f5f5f5;
}

.hidden {
    display: none;
}

.textinput {
    border-radius: 4px;
    width: 100%;
    color: $gray;
    font-size: 14px;
    font-weight: 300;
    border: 1px solid $light_gray;
    overflow: none;
    outline: none;
    -webkit-box-shadow: none;
    box-shadow: inset 1px 1px 2px 0 hsla(0, 0%, 0%, 0.08);
    padding: 10px 10px;
    -webkit-box-sizing: border-box;
    box-sizing: border-box;
    height: 37px;
    letter-spacing: 0.03em;
    line-height: 1.3em;

    &::-webkit-input-placeholder {
        color: $placeholder_color;
    }
    &::-moz-placeholder {
        color: $placeholder_color;
    }
    &:-ms-input-placeholder {
        color: $placeholder_color;
    }
    &:-moz-placeholder {
        color: $placeholder_color;
    }
}

textarea.textinput {
    height: 120px;
    padding: 10px;
    overflow: auto;
    resize: none;
    font-family: 'Roboto', sans-serif;
}
