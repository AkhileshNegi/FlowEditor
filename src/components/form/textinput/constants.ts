import { CompletionOption, getCompletionName } from '~/store/flowContext';

import * as xlnt from './functions.json';

// https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key/Key_Values
export enum KeyValues {
    KEY_AT = '@',
    KEY_SPACE = ' ',
    KEY_ENTER = 'Enter',
    KEY_UP = 'ArrowUp',
    KEY_DOWN = 'ArrowDown',
    KEY_RIGHT = 'ArrowRight',
    KEY_LEFT = 'ArrowLeft',
    KEY_TAB = 'Tab',
    KEY_ESC = 'Escape',
    KEY_BACKSPACE = 'Backspace',
    KEY_N = 'n',
    KEY_P = 'p',
    KEY_F = 'f',
    KEY_B = 'b'
}

export enum TopLevelVariables {
    contact = 'contact',
    input = 'input',
    results = 'results',
    run = 'run',
    parent = 'parent',
    child = 'child',
    trigger = 'trigger'
}

export const MAX_GSM_SINGLE = 160;
export const MAX_GSM_MULTI = 153;
export const MAX_UNICODE_SINGLE = 70;
export const MAX_UNICODE_MULTI = 67;

export const COMPLETION_HELP = 'Tab to complete, enter to select';

export const getFlowOptions = (accessor: string = ''): CompletionOption[] => {
    const prefix = accessor ? accessor : TopLevelVariables.run;
    return [
        { name: `${prefix}.flow`, summary: `The flow in which a ${accessor} run takes place` },
        {
            name: `${prefix}.flow.uuid`,
            summary: `The UUID of the flow in which a ${accessor} run takes place`
        },
        {
            name: `${prefix}.flow.name`,
            summary: `The name of the flow in which a ${accessor} run takes place`
        },
        {
            name: `${prefix}.flow.revision`,
            summary: `The revision number of the flow in which a ${accessor} run takes place`
        }
    ];
};

export const getWebhookOptions = (accessor: string = ''): CompletionOption[] => {
    const prefix = accessor ? accessor : TopLevelVariables.run;
    return [
        {
            name: `${prefix}.webhook`,
            summary: `The body of the response to the last webhook request made in a ${accessor} run`
        },
        {
            name: `${prefix}.webhook.status`,
            summary: `The status of the last webhook request made in a ${accessor} run`
        },
        {
            name: `${prefix}.webhook.status_code`,
            summary: `The status code returned from the last webhook request made in a ${accessor} run`
        },
        {
            name: `${prefix}.webhook.url`,
            summary: `The URL that was called by the last webhook request in a ${accessor} run`
        },
        {
            name: `${prefix}.webhook.body`,
            summary: `The body of the last webhook request made in a ${accessor} run`
        },
        {
            name: `${prefix}.webhook.json`,
            summary: `The JSON parsed body of the response to the last webhook request in a ${accessor} run, can access subelements`
        },
        {
            name: `${prefix}.webhook.request`,
            summary: `The raw request last made in a ${accessor} run, including headers`
        },
        {
            name: `${prefix}.webhook.response`,
            summary: `The raw response last received in a ${accessor} run, including headers`
        },
        { name: `${prefix}.results`, summary: `Results collected in a ${accessor} run` }
    ];
};

export const getInputOptions = (accessor: string = ''): CompletionOption[] => {
    const prefix = accessor ? accessor : TopLevelVariables.run;
    return [
        { name: `${prefix}.input`, summary: `A ${accessor} run's last input` },
        { name: `${prefix}.input.uuid`, summary: `The UUID of a ${accessor} run's last input` },
        { name: `${prefix}.input.type`, summary: `The type of a ${accessor} run's last input` },
        {
            name: `${prefix}.input.channel`,
            summary: `The channel a ${accessor} run's last input was received on`
        },
        {
            name: `${prefix}.input.created_on`,
            summary: `The time a ${accessor} run's last input was created`
        },
        {
            name: `${prefix}.input.text`,
            summary: `The text of a ${accessor} run's last message`
        },
        {
            name: `${prefix}.input.attachments`,
            summary: `The attachments on a ${accessor} run's last message`
        },
        {
            name: `${prefix}.input.urn`,
            summary: `The URN a ${accessor} run's last input was received on`
        }
    ];
};

export const getContactOptions = (accessor?: string): CompletionOption[] => {
    const prefix = accessor ? `${accessor}.` : '';
    const descriptor = accessor
        ? accessor === TopLevelVariables.run
            ? `${accessor}'s`
            : `${accessor} run's`
        : '';
    return [
        { name: `${prefix}contact`, summary: `The name of a ${descriptor} contact` },
        { name: `${prefix}contact.name`, summary: `The name of a ${descriptor} contact` },
        {
            name: `${prefix}contact.first_name`,
            summary: `The first name of a ${descriptor} contact`
        },
        {
            name: `${prefix}contact.language`,
            summary: `The language code for a ${descriptor} contact`
        },
        {
            name: `${prefix}contact.fields`,
            summary: `Custom fields on a ${descriptor} contact`
        },
        {
            name: `${prefix}contact.groups`,
            summary: `The groups a ${descriptor} contact is a member of`
        },
        { name: `${prefix}contact.urns`, summary: `URNs on a ${descriptor} contact` },
        {
            name: `${prefix}contact.urns.tel`,
            summary: `The preferred telephone number for a ${descriptor} contact`
        },
        {
            name: `${prefix}contact.urns.telegram`,
            summary: `The preferred telegram id for a ${descriptor} contact`
        },
        {
            name: `${prefix}contact.channel`,
            summary: `A ${descriptor} contact's preferred channel`
        },
        {
            name: `${prefix}contact.channel.uuid`,
            summary: `The UUID of a ${descriptor} contact's preferred channel`
        },
        {
            name: `${prefix}contact.channel.name`,
            summary: `The name of a ${descriptor} contact's preferred channel`
        },
        {
            name: `${prefix}contact.channel.address`,
            summary: `The address of a ${descriptor} contact's preferred channel`
        }
    ];
};

export const RUN_OPTIONS: CompletionOption[] = [
    { name: TopLevelVariables.run, summary: 'A run in this flow' },
    ...getFlowOptions(),
    ...getContactOptions(TopLevelVariables.run),
    ...getInputOptions(),
    ...getWebhookOptions()
];

export const CHILD_OPTIONS: CompletionOption[] = [
    { name: TopLevelVariables.child, summary: 'Run details collected in a child flow, if any' },
    ...getFlowOptions(TopLevelVariables.child),
    ...getContactOptions(TopLevelVariables.child),
    ...getInputOptions(TopLevelVariables.child),
    ...getWebhookOptions(TopLevelVariables.child)
];

export const PARENT_OPTIONS: CompletionOption[] = [
    {
        name: TopLevelVariables.parent,
        summary: 'Run details collected by a parent flow, if any'
    },
    ...getFlowOptions(TopLevelVariables.parent),
    ...getContactOptions(TopLevelVariables.parent),
    ...getInputOptions(TopLevelVariables.parent),
    ...getWebhookOptions(TopLevelVariables.parent)
];

export const TRIGGER_OPTIONS: CompletionOption[] = [
    { name: 'trigger', summary: 'A trigger that initiated a session' },
    { name: 'trigger.type', summary: 'The type of a trigger, one of “manual” or “flow”' },
    { name: 'trigger.params', summary: 'The parameters passed to a trigger' }
];

export const OPTIONS: CompletionOption[] = [
    ...getContactOptions(),
    { name: 'results', summary: 'The results for the current run' },
    ...RUN_OPTIONS,
    ...CHILD_OPTIONS,
    ...PARENT_OPTIONS,
    ...TRIGGER_OPTIONS,
    ...xlnt
];

export const TOP_LEVEL_OPTIONS = OPTIONS.filter((option: CompletionOption) => {
    const name = getCompletionName(option);
    return (
        name === TopLevelVariables.contact ||
        name === TopLevelVariables.input ||
        name === TopLevelVariables.results ||
        name === TopLevelVariables.run ||
        name === TopLevelVariables.parent ||
        name === TopLevelVariables.child ||
        name === TopLevelVariables.trigger
    );
});

export const GSM: { [key: string]: number } = {
    // char: charCode
    0: 48,
    1: 49,
    2: 50,
    3: 51,
    4: 52,
    5: 53,
    6: 54,
    7: 55,
    8: 56,
    9: 57,
    '\n': 10,
    '\f': 12,
    '\r': 13,
    ' ': 32,
    '!': 33,
    '"': 34,
    '#': 35,
    $: 36,
    '%': 37,
    '&': 38,
    "'": 39,
    '(': 40,
    ')': 41,
    '*': 42,
    '+': 43,
    ',': 44,
    '-': 45,
    '.': 46,
    '/': 47,
    ':': 58,
    ';': 59,
    '<': 60,
    '=': 61,
    '>': 62,
    '?': 63,
    '@': 64,
    A: 65,
    B: 66,
    C: 67,
    D: 68,
    E: 69,
    F: 70,
    G: 71,
    H: 72,
    I: 73,
    J: 74,
    K: 75,
    L: 76,
    M: 77,
    N: 78,
    O: 79,
    P: 80,
    Q: 81,
    R: 82,
    S: 83,
    T: 84,
    U: 85,
    V: 86,
    W: 87,
    X: 88,
    Y: 89,
    Z: 90,
    '[': 91,
    '\\': 92,
    ']': 93,
    '^': 94,
    _: 95,
    a: 97,
    b: 98,
    c: 99,
    d: 100,
    e: 101,
    f: 102,
    g: 103,
    h: 104,
    i: 105,
    j: 106,
    k: 107,
    l: 108,
    m: 109,
    n: 110,
    o: 111,
    p: 112,
    q: 113,
    r: 114,
    s: 115,
    t: 116,
    u: 117,
    v: 118,
    w: 119,
    x: 120,
    y: 121,
    z: 122,
    '{': 123,
    '|': 124,
    '}': 125,
    '~': 126,
    '¡': 161,
    '£': 163,
    '¤': 164,
    '¥': 165,
    '§': 167,
    '¿': 191,
    Ä: 196,
    Å: 197,
    Æ: 198,
    Ç: 199,
    É: 201,
    Ñ: 209,
    Ö: 214,
    Ø: 216,
    Ü: 220,
    ß: 223,
    à: 224,
    ä: 228,
    å: 229,
    æ: 230,
    è: 232,
    é: 233,
    ì: 236,
    ñ: 241,
    ò: 242,
    ö: 246,
    ø: 248,
    ù: 249,
    ü: 252,
    Γ: 915,
    Δ: 916,
    Θ: 920,
    Λ: 923,
    Ξ: 926,
    Π: 928,
    Σ: 931,
    Φ: 934,
    Ψ: 936,
    Ω: 937,
    '€': 8364
};
