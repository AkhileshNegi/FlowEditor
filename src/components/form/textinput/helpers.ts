import { split } from 'split-sms';
import { GSM, OPTIONS } from '~/components/form/textinput/constants';
import { AssetMap, AssetStore, CompletionOption } from '~/store/flowContext';

export interface UnicodeCharMap {
    [char: string]: boolean;
}

interface MsgStats {
    value: string;
    parts: string[];
    characterCount: number;
    unicodeChars: UnicodeCharMap;
}

export const isUnicode = (char: string): boolean => {
    if (GSM.hasOwnProperty(char)) {
        return false;
    }
    return true;
};

export const getUnicodeChars = (msg: string): UnicodeCharMap => {
    const chars = {};

    for (const char of msg) {
        if (isUnicode(char)) {
            chars[char] = true;
        }
    }

    return chars;
};

/**
 * Replaces unicode characters commonly inserted by text editors like MSWord with their GSM equivalents
 * @param {string} msg - msg to be cleaned
 * @returns {string} Cleaned msg
 */
export const cleanMsg = (msg: string): string =>
    msg
        .replace(/[\u2018\u2019]/g, "'") // Smart single quotes
        .replace(/[\u201C\u201D]/g, '"') // Smart double quotes
        .replace(/[\u2013\u2014]/g, '-') // En/em dash
        .replace(/\u2026/g, '...') // Horizontal ellipsis
        .replace(/\u2002/g, ' '); // En space

/**
 * First pass at providing the user with an accurate character count for their SMS messages.
 * Determines encoding, segments, max character limit per message and calculates character count.
 * Optionally replaces common unicode 'gotcha characters' with their GSM counterparts.
 * @param value
 * @param replace
 */
export const getMsgStats = (value: string | string[], replace?: boolean): MsgStats => {
    let newVal = value as string;

    // Localized values are stored as string arrays
    if (newVal.constructor === Array) {
        newVal = newVal[0];
    }

    if (replace) {
        newVal = cleanMsg(newVal);
    }

    const stats = split(newVal);

    return {
        value: newVal,
        parts: stats.parts,
        characterCount: stats.length,
        unicodeChars: getUnicodeChars(newVal)
    };
};

export const filterOptions = (
    options: CompletionOption[],
    query: string = '',
    includeFunctions: boolean
): CompletionOption[] => {
    const search = query.toLowerCase();
    return options.filter((option: CompletionOption) => {
        if (includeFunctions) {
            if (option.signature) {
                return option.signature.indexOf(search) === 0;
            }
        }

        if (option.name) {
            const rest = option.name.substr(search.length);
            return (
                option.name.indexOf(search) === 0 &&
                (rest.length === 0 || rest.substr(1).indexOf('.') === -1)
            );
        }
    });
};

export const getResultPropertyOptions = (accessor: string, name: string) => [
    {
        name: accessor,
        description: `Result for "${name}"`
    },
    {
        name: `${accessor}.value`,
        description: `Value for "${name}"`
    },
    {
        name: `${accessor}.category`,
        description: `Category for "${name}"`
    },
    {
        name: `${accessor}.category_localized`,
        description: `Localized category for "${name}"`
    },
    {
        name: `${accessor}.input`,
        description: `Input for "${name}"`
    },
    {
        name: `${accessor}.created_on`,
        description: `Time "${name}" was created`
    }
];

/* export const getResultOptions = (results: ResultMap) =>
    [...new Set(Object.keys(results).map(uuid => results[uuid]))].reduce((options, query) => {
        const accessor = query.replace(/^@/, '');
        const name = titleCase(accessor.slice(accessor.lastIndexOf('.') + 1).replace(/_/g, ' '));
        options.push(...getResultPropertyOptions(accessor, name));
        return options;
    }, []);
*/

export const getContactFieldOptions = (assets: AssetMap) =>
    Object.keys(assets).reduce((options, key) => {
        const { [key]: asset } = assets;
        options.push({
            name: `fields.${key}`,
            summary: `${asset.name} for the contact.`
        });

        const accessors = ['', 'parent.', 'run.', 'child.'];
        accessors.forEach(accessor =>
            options.push({
                name: `${accessor}contact.fields.${key}`,
                summary: `${asset.name} for the contact.`
            })
        );
        return options;
    }, []);

export const getResultsOptions = (assets: AssetMap) =>
    Object.keys(assets).reduce((options, key) => {
        const { [key]: asset } = assets;
        const accessors = ['results', 'run.results'];
        accessors.forEach(accessor => {
            options.push({
                name: `${accessor}.${key}`,
                summary: `${asset.name} for the run.`
            });

            options.push({
                name: `${accessor}.${key}.category`,
                summary: `${asset.name} category for the run.`
            });

            options.push({
                name: `${accessor}.${key}.category_localized`,
                summary: `${asset.name} localized category for the run.`
            });
        });

        return options;
    }, []);

export const getOptionsList = (autocomplete: boolean, assets: AssetStore): CompletionOption[] => {
    return autocomplete
        ? [
              ...OPTIONS,
              ...getContactFieldOptions(assets.fields ? assets.fields.items : {}),
              ...getResultsOptions(assets.results ? assets.results.items : {})
          ]
        : OPTIONS;
};

export const pluralize = (count: number, noun: string, suffix: string = 's'): string =>
    `${noun}${count !== 1 ? suffix : ''}`;
