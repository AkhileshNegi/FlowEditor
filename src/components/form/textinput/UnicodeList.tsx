import { UnicodeCharMap } from 'components/form/textinput/helpers';
import * as React from 'react';
import { pluralize, renderIf } from 'utils';

import styles from './UnicodeList.module.scss';

export interface UnicodeListProps {
  unicodeChars: UnicodeCharMap;
}

export const utfWarning = (num: number) =>
  ` Note that your message may require more segments to send because it contains the following ${pluralize(
    num,
    'character'
  )}:`;

export const unicodeListContainerSpecId = 'unicode-list-container';
export const unicodeListSpecId = 'unicode-list';

const UnicodeList: React.SFC<UnicodeListProps> = ({ unicodeChars }) => {
  const chars = Object.keys(unicodeChars);
  return renderIf(chars.length > 0)(
    <span data-testid={unicodeListContainerSpecId}>
      {utfWarning(chars.length)}
      <div className={styles.unicode_list} data-testid={unicodeListSpecId}>
        {chars.map((char, idx) => (
          <span key={char} className={styles.unicode_char}>
            {char}
          </span>
        ))}
      </div>
    </span>
  );
};

export default UnicodeList;
