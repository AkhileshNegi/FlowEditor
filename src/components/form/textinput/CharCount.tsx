import * as React from 'react';
import * as classNames from 'classnames/bind';
import * as styles from '~/components/form/textinput/CharCount.scss';
import { UnicodeCharMap, pluralize } from '~/components/form/textinput/helpers';
import UnicodeList from '~/components/form/textinput/UnicodeList';

interface CharCountProps {
    count: number;
    parts: number;
    unicodeChars: UnicodeCharMap;
}

const cx = classNames.bind(styles);

const CharCount: React.SFC<CharCountProps> = ({ count, parts, unicodeChars }) => {
    const hasUnicode = Object.keys(unicodeChars).length > 0;
    const toolTipTextClasses = cx({
        [styles.tooltipText]: true,
        [styles.unicode]: hasUnicode
    });
    return (
        <div className={styles.count}>
            <div className={styles.tooltip}>
                {count}
                <span className={styles.divider}>/</span>
                {parts}
                <span className={toolTipTextClasses}>
                    This message will be about <b>{count}</b> {pluralize(count, 'character')} long.
                    We estimate it will take <b>{parts}</b> {pluralize(parts, 'segment')} to send
                    over SMS.
                    {hasUnicode && <UnicodeList unicodeChars={unicodeChars} />}
                </span>
            </div>
        </div>
    );
};

export default CharCount;
