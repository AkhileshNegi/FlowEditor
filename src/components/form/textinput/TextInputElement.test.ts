import setCaretPosition from 'get-input-selection';
import { KeyValues } from '~/components/form/textinput/constants';
import {
    TextInputElement,
    TextInputProps,
    TextInputState
} from '~/components/form/textinput/TextInputElement';
import { getTypeConfig, Types } from '~/config/typeConfigs';
import { composeComponentTestUtils } from '~/testUtils';

// import { wrapper } from '~/components/form/textinput/TextInputElement.scss';
const baseProps: TextInputProps = {
    name: 'Message',
    typeConfig: getTypeConfig(Types.send_msg),
    assetStore: {}
};

const { setup, spyOn } = composeComponentTestUtils(TextInputElement, baseProps);

jest.mock('get-input-selection', () => ({
    default: jest.fn()
}));

const mockEvent = {
    preventDefault(): void {
        return;
    },
    stopPropagation(): void {
        return;
    }
};

const simulateString = (wrap: any, keys: string) => {
    const input = wrap.find('textarea');
    let presses = wrap.instance().textEl.value;
    for (const key of keys) {
        input.prop('onKeyDown')({
            ...mockEvent,
            key
        });

        presses += key;

        input.prop('onChange')({
            currentTarget: {
                value: presses,
                selectionStart: presses.length
            }
        });
    }
};

const simulateKey = (wrap: any, key: KeyValues, ctrlKey: boolean = false) => {
    if (key === KeyValues.KEY_AT) {
        simulateString(wrap, key);
    } else {
        const input = wrap.find('textarea');
        input.prop('onKeyDown')({
            ...mockEvent,
            key,
            ctrlKey
        });
    }
};

const getState = (wrap: any): TextInputState => {
    return wrap.instance().state as TextInputState;
};

const createWrapper = () => {
    return setup(false, {
        $merge: {
            onChange: jest.fn(),
            textarea: true,
            autocomplete: true
        }
    }).wrapper;
};

describe(TextInputElement.name, () => {
    afterEach(() => {
        setCaretPosition.mockReset();
    });

    describe('function context', () => {
        it('should use the most recent incomplete function', () => {
            const wrapper = createWrapper();
            simulateString(wrapper, '@(max(default(contact.rank, cont');

            const state = getState(wrapper);
            expect(state.completionVisible).toBeTruthy();
            expect(state.query).toEqual('cont');
            expect(state.matches.length).toBe(1);
            expect(state.fn.signature).toEqual('default(value, default)');
        });

        it('should use fold back on function completion', () => {
            const wrapper = createWrapper();
            simulateString(wrapper, '@(max(default(contact.rank, contact.age), ');

            const state = getState(wrapper);
            expect(state.query).toEqual('');
            expect(state.fn.signature).toEqual('max(values...)');
        });
    });

    describe('filtering', () => {
        it('should bring up completion menu for top level options', () => {
            const wrapper = createWrapper();
            simulateString(wrapper, '@');

            const state = getState(wrapper);
            expect(state.completionVisible).toBeTruthy();
            expect(state.query).toEqual('');
            expect(state.caretOffset).toEqual(1);

            // should show all top level options
            expect(state.matches.length).toBe(5);
        });

        it('should show filter options', () => {
            const wrapper = createWrapper();
            simulateString(wrapper, '@co');

            const state = getState(wrapper);
            expect(state.completionVisible).toBeTruthy();
            expect(state.query).toEqual('co');
            expect(state.caretOffset).toEqual(3);

            // only our contact option should be there
            expect(state.matches.length).toBe(1);
        });

        it('should not match functions without open paren', () => {
            const wrapper = createWrapper();
            simulateString(wrapper, '@max');

            const state = getState(wrapper);
            expect(state.matches.length).toBe(0);
            expect(state.completionVisible).toBeFalsy();
        });

        it('should bring up completion menu for top level options and functions', () => {
            const wrapper = createWrapper();
            simulateString(wrapper, '@(');

            const state = getState(wrapper);
            // should show all top level options and functions
            expect(state.matches.length).toBeGreaterThan(60);
            expect(state.completionVisible).toBeTruthy();
        });

        it('should bring up functions within text', () => {
            const wrapper = createWrapper();
            simulateString(wrapper, 'some text @(');

            const state = getState(wrapper);
            // should show all top level options and functions
            expect(state.matches.length).toBeGreaterThan(60);
            expect(state.completionVisible).toBeTruthy();
        });
    });

    describe('navigation', () => {
        it('should filter after tabbing', () => {
            const wrapper = createWrapper();
            simulateString(wrapper, '@cont');
            simulateKey(wrapper, KeyValues.KEY_TAB);
            simulateString(wrapper, '.fir');

            const state = getState(wrapper);
            expect(state.completionVisible).toBeTruthy();
            expect(state.query).toEqual('contact.fir');
            expect(state.caretOffset).toEqual(12);

            // tabbing forward should give us all our contact options
            expect(state.matches.length).toBe(1);
            expect(state.matches[0].name).toBe('contact.first_name');
        });

        it('should handle completion w/ "Enter" key', () => {
            const wrapper = createWrapper();
            simulateString(wrapper, '@cont');
            simulateKey(wrapper, KeyValues.KEY_ENTER);

            const state = getState(wrapper);
            expect(state.completionVisible).toBeFalsy();
            expect(state.query).toEqual('');
            expect(state.caretOffset).toEqual(8);
            expect(state.matches.length).toBe(0);
        });

        it('should allow navigation with arrow keys', () => {
            const wrapper = createWrapper();
            simulateString(wrapper, '@c');

            simulateKey(wrapper, KeyValues.KEY_DOWN);
            let state = getState(wrapper);
            expect(state.selectedOptionIndex).toBe(1);

            simulateKey(wrapper, KeyValues.KEY_UP);
            state = getState(wrapper);
            expect(state.selectedOptionIndex).toBe(0);
        });

        it('should allow navigation with OSX shortcuts', () => {
            const wrapper = createWrapper();
            simulateString(wrapper, '@c');

            simulateKey(wrapper, KeyValues.KEY_N, true);
            let state = getState(wrapper);
            expect(state.selectedOptionIndex).toBe(1);

            simulateKey(wrapper, KeyValues.KEY_P, true);
            state = getState(wrapper);
            expect(state.selectedOptionIndex).toBe(0);
        });

        it('should handle completion w/ "Tab" key', () => {
            const wrapper = createWrapper();
            simulateString(wrapper, '@cont');
            simulateKey(wrapper, KeyValues.KEY_TAB);

            const state = getState(wrapper);
            expect(state.completionVisible).toBeTruthy();
            expect(state.query).toEqual('contact');
            expect(state.caretOffset).toEqual(8);

            // tabbing forward should give us all our contact options
            expect(state.matches.length).toBe(8);
        });
    });

    describe('visibility', () => {
        it('should hide if outside completed expression', () => {
            const wrapper = createWrapper();
            simulateString(wrapper, '@(max(contact.first, contact.second)) ');

            const state = getState(wrapper);
            expect(state.completionVisible).toBeFalsy();
            expect(state.matches.length).toBe(0);
        });
    });
});
