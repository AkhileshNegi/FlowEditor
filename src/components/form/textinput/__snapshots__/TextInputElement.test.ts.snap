// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInputElement completion should complete in the middle of text 1`] = `
Array [
  Object {
    "name": "contact",
    "summary": "The name of the  contact",
  },
  Object {
    "name": "contact.first_name",
    "summary": "The first name of a  contact",
  },
  Object {
    "name": "contact.name",
    "summary": "The name of the  contact",
  },
  Object {
    "name": "contact.display",
    "summary": "The name or urn of the  contact",
  },
  Object {
    "name": "contact.fields",
    "summary": "Custom fields on a  contact",
  },
  Object {
    "name": "contact.urn",
    "summary": "The primary urn for a  contact",
  },
  Object {
    "name": "contact.language",
    "summary": "The language iso code for a  contact",
  },
  Object {
    "name": "contact.created_on",
    "summary": "The creation date for the  contact",
  },
  Object {
    "name": "contact.groups",
    "summary": "The groups a  contact is a member of",
  },
  Object {
    "name": "contact.urns",
    "summary": "The urns for a  contact",
  },
  Object {
    "name": "contact.uuid",
    "summary": "The uuid of the  contact",
  },
  Object {
    "name": "contact.channel",
    "summary": "A  contact's preferred channel",
  },
]
`;

exports[`TextInputElement completion should handle completion w/ "Tab" key 1`] = `
Array [
  Object {
    "name": "contact",
    "summary": "The name of the  contact",
  },
  Object {
    "name": "contact.first_name",
    "summary": "The first name of a  contact",
  },
  Object {
    "name": "contact.name",
    "summary": "The name of the  contact",
  },
  Object {
    "name": "contact.display",
    "summary": "The name or urn of the  contact",
  },
  Object {
    "name": "contact.fields",
    "summary": "Custom fields on a  contact",
  },
  Object {
    "name": "contact.urn",
    "summary": "The primary urn for a  contact",
  },
  Object {
    "name": "contact.language",
    "summary": "The language iso code for a  contact",
  },
  Object {
    "name": "contact.created_on",
    "summary": "The creation date for the  contact",
  },
  Object {
    "name": "contact.groups",
    "summary": "The groups a  contact is a member of",
  },
  Object {
    "name": "contact.urns",
    "summary": "The urns for a  contact",
  },
  Object {
    "name": "contact.uuid",
    "summary": "The uuid of the  contact",
  },
  Object {
    "name": "contact.channel",
    "summary": "A  contact's preferred channel",
  },
]
`;

exports[`TextInputElement filtering should bring up completion menu for top level options 1`] = `
Array [
  Object {
    "name": "contact",
    "summary": "The name of the  contact",
  },
  Object {
    "name": "results",
    "summary": "The results for the current run",
  },
  Object {
    "name": "fields",
    "summary": "The custom fields for the contact",
  },
  Object {
    "name": "input",
    "summary": "The most recent input",
  },
  Object {
    "name": "urns",
    "summary": "The urns for the contact",
  },
  Object {
    "name": "run",
    "summary": "A run in this flow",
  },
  Object {
    "name": "child",
    "summary": "Run details collected in a child flow, if any",
  },
  Object {
    "name": "parent",
    "summary": "Run details collected by a parent flow, if any",
  },
  Object {
    "name": "webhook",
    "summary": "The parsed JSON payload of the last webhook call",
  },
  Object {
    "name": "trigger",
    "summary": "A trigger that initiated a session",
  },
]
`;

exports[`TextInputElement filtering should show filter at the second level 1`] = `
Array [
  Object {
    "name": "contact.first_name",
    "summary": "The first name of a  contact",
  },
]
`;

exports[`TextInputElement filtering should show filter options 1`] = `
Array [
  Object {
    "name": "contact",
    "summary": "The name of the  contact",
  },
]
`;

exports[`TextInputElement function context should use the most recent incomplete function 1`] = `
Array [
  Object {
    "name": "contact",
    "summary": "The name of the  contact",
  },
]
`;

exports[`TextInputElement navigation should filter after tabbing 1`] = `
Array [
  Object {
    "name": "contact.first_name",
    "summary": "The first name of a  contact",
  },
]
`;
